### ISSUE TITLE
how to deal with require hackery

### ISSUE THREAD

**Issue by faassen** [jspm/jspm-cli#441]
I just tried to use the pretender library with jspm.

This fails -- it tries a request for ``/base/jspm_packages/npm/pretender@0.4.2/bower_components/FakeXMLHttpRequest/fake_xml_http_request.js`` but no such ``bower_components directory`` exists.

This is due to some hackery at the start of this file:

https://github.com/trek/pretender/blob/master/pretender.js

First it tries to detect whether it's in a node evironment and if so does a hardcoded import from ``bower_components``.

Clearly this isn't going to work. The question is how we help library authors to rewrite code like this so it does work with jspm. Better documentation?

One bit I don't understand is that it tries to detect whether things are running in node or the browser, and then the code (even though I have it running in Chrome through karma) thinks it's in node. If it thought it was in the browser things might very well have worked -- not sure.

See also https://github.com/trek/pretender/issues/54

**Comment by guybedford** [comment]
The require hackery is not actually the issue here - the npm endpoint automatically defines the `process` global correctly.

Rather this is a more fundamental issue which does require workarounds - because requires are statically analyzed in the loader (to support an async require pipeline), all requires are picked up by jspm.

So even though the code path never triggers, it still tries to have the code loaded already in case it does to provide it sync.

To work around this, try doing:

```
  jspm install npm:pretender -o "{ map: { './bower_components/FakeXMLHttpRequest/fake_xml_http_request': '@empty' } }"
```

**Comment by guybedford** [comment]
The above can be stored as an override in the registry to make this work for others, or added to the package.json of the original repo (ideally probably both).

**Comment by guybedford** [comment]
I've added the above override into the registry now - so this should work with a simple:

```
  jspm install npm:pretender
```

Let me know how it goes.

**Comment by faassen** [comment]
Thanks! I started out with sinon, but didn't get it to work. Then tried pretender. Then went back to sinon using a karma plugin that made stuff work again. I'll try pretender again -- it has a few features that I like.

**Comment by faassen** [comment]
I just tried this, but something isn't right yet I think. I tried:

```
import Pretender from 'pretender'
```

but ``Pretender`` is an object, not a function to create the pretender class as I expected. It's possible my expectation is wrong though - I haven't tried pretender outside of jspm.

**Comment by guybedford** [comment]
@faassen some hackery does come into play here unfortunately. This is entirely because Pretender isn't behaving like an npm package, but rather like a bower package - that is it is specially designing its `isNode` function not to trigger with Browserify's process library so that it always reverts to global variables even in Browserify use, which is not ideal.

To make it work, I'd suggest manually shimming it like a global library:

```
jspm install npm:pretender -o "{ shim: { pretender: { deps: ['route-recognizer', 'fake-xml-http-request'], exports: 'Pretender' } } }
```

Please try the above install command and let me know if that works out for you ok. If so we can add the above to the override so that others can benefit from this work.

**Comment by guybedford** [comment]
@faassen I've added pretender into the registry with https://github.com/jspm/registry/commit/83d02d35f47dc1c20044482d23b4be9dd0b253d5 and https://github.com/jspm/registry/commit/517c7e90d87cf5de350b847cd73957bf0c3a8474.

Route Recognizer was running as a CommonJS module, and not defining window.RouteRecognizer needed by pretender, so I added an override for this as well to be treated like a global script instead of CommonJS - https://github.com/jspm/registry/commit/9b28842a46badfe6ea5c4e73fdbfbbf7cf602675.

With these,

```
  jspm install pretender
```

now works entirely as expected.

This is the painful part of jspm where we have to fix the anti-modularity patterns already out there. The point is we have the tools to do this with overrides, and expand the working jspm libraries in the process.
